File: java/40.combination-sum-ii.java
import java.util.List;

/*
 * @lc app=leetcode id=40 lang=java
 *
 * [40] Combination Sum II
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        if (candidates.length == 0) return res;
        Arrays.sort(candidates);
        backtrack(0, candidates, target, new ArrayList<>());
        return res;
    }

    private void backtrack(int start, int[] candidates, int target, List<Integer> tmpList) {
        if (target == 0) {
            res.add(new ArrayList<>(tmpList));
            return;
        }
        int prev = -1;
        for (int i = start; i < candidates.length && candidates[i] <= target; ++i) {
            if (candidates[i] == prev) {
                continue;
            }
            prev = candidates[i];
            tmpList.add(candidates[i]);
            backtrack(i+1, candidates, target-candidates[i], tmpList);
            tmpList.remove(tmpList.size()-1);

        }
    }

}
// @lc code=end

File: go/40.combination-sum-ii.go
/*
 * @lc app=leetcode id=40 lang=golang
 *
 * [40] Combination Sum II
 */

// @lc code=start
func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	res := [][]int{}
	var backtrack func(int, int, []int)
	backtrack = func(idx, acc int, comb []int) {
		if acc == target {
			tmp := make([]int, len(comb))
			copy(tmp, comb)
			res = append(res, tmp)
			return
		}
		for i := idx; i < len(candidates); {
			cur := candidates[i]
			if acc + cur > target {
				break
			}
			ni := i+1
			for ni < len(candidates) && candidates[ni] == cur {
				ni++
			}
			originsize, acc2 := len(comb), acc
			for j := i; j < ni; j++ {
				acc2 += cur
				if acc2 > target {
					break
				}
				comb = append(comb, cur)
				backtrack(ni, acc2, comb)
			}
			comb = comb[:originsize]
			i = ni
		}
	}
	
	backtrack(0, 0, []int{})
	return res
}
// @lc code=end

